### PR TITLE
KAFKA-18629: Delete share group state impl [1/N]

### PR DESCRIPTION
**Pull request by smjn**
* In this PR, we have added the impl for supporting `DeleteShareGroupState` API requests.
* This is the first in a series of PRs for end to end impl of the same.
* This PR has KafkaApis, ShareCoordinatorService and ShareCoordinatorShard impls along with requisite tests.
* ref: https://cwiki.apache.org/confluence/display/KAFKA/KIP-932%3A+Queues+for+Kafka#KIP932:QueuesforKafka-DeleteShareGroupStateAPI

### CODE REVIEW

File: share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorShardTest.java [hunk 1]
```diff
@@ -957,6 +961,144 @@ public void testReadStateLeaderEpochUpdateNoUpdate() {
...
+        assertEquals(expectedData, result.response());
+        assertEquals(expectedRecords, result.records());
+
+        assertNull(shard.getShareStateMapValue(shareCoordinatorKey));
+        assertNull(shard.getLeaderMapValue(shareCoordinatorKey));
+        assertNull(shard.getStateEpochMapValue(shareCoordinatorKey));
+    }
+
+    @Test
+    public void testDeleteStateInvalidRequestData() {
```
Comment (clolov) [inline share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorShardTest.java:1060]: Is it easy for you to also test the condition `topicId == null`? Or is it tested elsewhere and I have missed it?

File: share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorShardTest.java [hunk 2]
```diff
@@ -957,6 +961,144 @@ public void testReadStateLeaderEpochUpdateNoUpdate() {
...
+        List<CoordinatorRecord> expectedRecords = List.of();
+
+        assertEquals(expectedData, result.response());
+        assertEquals(expectedRecords, result.records());
+    }
+
+    @Test
+    public void testDeleteNullMetadataImage() {
+        ShareCoordinatorShard shard = new ShareCoordinatorShardBuilder().build();
+        shard.onNewMetadataImage(null, null);
```
Comment (clolov) [inline share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorShardTest.java:1087]: As far as I can tell this tests the case where the `metadataImage == null`, is it easy for you to test the `metadataImage.topics().getTopic(topicId) == null || metadataImage.topics().getPartition(topicId, partitionId) == null` condition as well?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala [hunk 3]
```diff
@@ -10513,6 +10513,96 @@ class KafkaApisTest extends Logging {
     })
   }
 
+  @Test
```
Comment (clolov) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:10516]: Is it easy to test the case where the share coordinator is not enabled? Or is this exercised someplace else which I have missed?

**Review by clolov — COMMENTED** [review]
Thanks for this! I am still catching up, so apologies if some of the questions don't make too much sense. I left a couple of comments here and there, but I had a more generic question. Is there something preventing a retried request to result in multiple tombstones? If not, is this problematic?

File: share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorShardTest.java (hunk 1 above)
Comment (smjn) [reply share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorShardTest.java:1060]: At code level it is not possible to pass in the topicId as null since the the call flow is:
DefaultStatePersister -> PersisterStateManager -> KafkaApis -> ShareCoordinatorService -> ShareCoordinatorShard.

The DefaultStatePersister is an implementation of the Persister interface which is used to communicate with the share coordinator via kafkaApis. This way callers are completely oblivious of the share coordinator. It can be easily swapped for some other abstraction.

Furthermore, this is an inter broker RPC which means external clients cannot directly invoke it via kafkaAdminClient. The topicId is null protected in PersisterStateManager.

For this new DeleteShareGroupState - the PersisterStateManager code has not been added yet. But we will be creating a new inner class over there which will extend PersisterStateManager.PersisterStateManagerHandler whose constructor takes care of doing the null check and preventing this scenario. Hence, it is not explicitly tested.

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala (hunk 3 above)
Comment (smjn) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:10516]: If share coordinator is not enabled - this call can never be made since this is an inter broker RPC and share coordinator being unavailable means the Persister implementation will never default to DefaultStatePersister and will be a NoOpStatePersister thus, this scenario does not come up. Added for completeness.

**Comment by smjn** [comment]
> Thanks for this! I am still catching up, so apologies if some of the questions don't make too much sense. I left a couple of comments here and there, but I had a more generic question. Is there something preventing a retried request to result in multiple tombstones? If not, is this problematic?

If you mean not being able to write to the log by the replica manager, then the coordinator runtime manages that. If it is unable to commit the tombstone - the CoordinatorRuntime.scheduleWriteOperation should eventually fail.

The caller side retry logic is encapsulated in the PersisterStateManager which only retries on certain well known error codes. These calls are not meant to be made directly via the client.

So consider the following points in the call flow

- CoordinatorRuntime (guarantee that record is written or failed)
- KafkaApis (response might be lost) -> but the persister state manager only retries on well known errors and NOT on TimeoutException.


Lets consider the worst case - If multiple tombstones do get written such that - 
```
INITIAL_TOMBSTONE_share_part_key_1
valid_record_1_share_part_key_1
valid_record_2_share_part_key_2
BAD_RETRY_NEW_TOMBSTONE_share_part_key_1
```

On replay of `BAD_RETRY_NEW_TOMBSTONE_share_part_key_1`. The state of `share_part_key_1` will be cleared in the ShareCoordinatorShard. The state is stored in TimelineHashMaps encapsulated in ShareCoordinatorShard. 
Later, when a read call happens this lost information will not be returned and as a result the ShareParitition will re-acquire the records and resend them. This by nature falls within the realm of at least once semantics.

**Review by AndrewJSchofield — APPROVED** [review]
Thanks for the PR. Looks good to me.
